predicators/nsrt_learning/strips_learning/effect_search_learner.py

```diff
@@ -21,160 +21,121 @@
-        if uncovered_transition is not None:
-            param_option, option_objs, add_effs, keep_effs = \
-                uncovered_transition
-            option_spec, lifted_add_effs, lifted_keep_effs = \
-                self._create_new_effect_set(param_option, option_objs, \
-                    add_effs, keep_effs)
-            new_effect_sets = curr_effect_sets.add(option_spec,
-                                                   lifted_add_effs,
-                                                   lifted_keep_effs)
-
-        return new_effect_sets
-
-    def get_successors(self,
-                       effect_sets: _EffectSets) -> Iterator[_EffectSets]:
-        initial_heuristic_val = self._associated_heuristic(effect_sets)
-        new_effect_sets = self._get_new_effect_sets_by_backchaining(
-            effect_sets)
-        if initial_heuristic_val > 0:
-            new_heuristic_val = self._associated_heuristic(new_effect_sets)
-            if new_heuristic_val == initial_heuristic_val:
-                # This means there was a keep effect problem with the new add
-                # effects we just induced. We need to call backchaining again
-                # to fix this.
-                new_effect_sets = self._get_new_effect_sets_by_backchaining(
-                    new_effect_sets)
-                new_heuristic_val = self._associated_heuristic(new_effect_sets)
-                assert new_heuristic_val < initial_heuristic_val
-        yield new_effect_sets
-
-    def _create_new_effect_set(
-        self, param_option: ParameterizedOption, option_objs: Sequence[Object],
-        add_effs: Set[GroundAtom], keep_effs: Set[GroundAtom]
-    ) -> Tuple[OptionSpec, Set[LiftedAtom], Set[LiftedAtom]]:
-        # Create a new effect set.
-        all_objs = sorted(
-            set(option_objs) | {o
-                                for a in add_effs for o in a.objects})
-        all_types = [o.type for o in all_objs]
-        all_vars = utils.create_new_variables(all_types)
-        obj_to_var = dict(zip(all_objs, all_vars))
-        option_vars = [obj_to_var[o] for o in option_objs]
-        option_spec = (param_option, option_vars)
-        lifted_add_effs = {a.lift(obj_to_var) for a in add_effs}
-        lifted_keep_effs = {a.lift(obj_to_var) for a in keep_effs}
-        return (option_spec, lifted_add_effs, lifted_keep_effs)
-
-    def _get_first_uncovered_transition(
-        self,
-        pnads: List[PartialNSRTAndDatastore],
-    ) -> Optional[Tuple[ParameterizedOption, Sequence[Object], Set[GroundAtom],
-                        Set[GroundAtom]]]:
-        # Find the first uncovered segment. Do this in a kind of breadth-first
-        # backward search over trajectories.
-        # Compute all the chains once up front.
+    def get_successors(
+        self, pnads: FrozenSet[PartialNSRTAndDatastore]
+    ) -> Iterator[FrozenSet[PartialNSRTAndDatastore]]:
+        pnads_list = list(pnads)
```

Comment:
the order of the pnads will be nondeterministic in this conversion from a set to a list. if you need a list, make sure it's sorted somehow. but you might not actually need a list?